### PR TITLE
Enhance discussion about routers and route specifications

### PR DESCRIPTION
**Pull request by hlship**

### CODE REVIEW

File: content/reference/important-protocols.adoc [hunk 1]
```diff
@@ -40,49 +44,61 @@ reference] for details of their behaviors.
...
 extension points and alternative strategies in several places.
 
-To give you an idea what that means, there are three built-in route
-definition syntaxes: table, terse, and verbose. There are three built
-in routing algorithms: linear-search, prefix-tree, and map-tree. An
-application can use any syntax with any algorithm. Or, an application
-can create a new syntax while using a built-in algorithm. Or, an
-application could use a built-in syntax with a new algorithm!
+Over time, Pedestal has evolved several data structures for representing routing;
+at the core, they map an HTTP method and URI path to a set of interceptors and a handler.
```
Comment (aredington) [inline content/reference/important-protocols.adoc:48]: May be worth mentioning that these data structures are tightly coupled to HTTP. Namespaces should suggest it, but may bear repeating.

**Review by aredington — APPROVED** [review]
This is great.

Worth considering highlighting HTTP centricity of routing considering we're also discussing how agnostic of HTTP interceptors are.

File: content/reference/chain-providers.adoc [hunk 2]
```diff
@@ -8,16 +8,22 @@ Cognitect
...
 
-For HTTP, the link:servlet-interceptor[servlet interceptor] is the
-chain provider.
+- Processing messages from a queue, such as Kafka or JMS
+- Performing a series of transformations on documents as part of a publishing service
+- Processing outgoing requests and incoming responses from a back-end service
+
+A _chain provider_ is responsible for accepting an incoming stimuli (such as an incoming HTTP request)
+and setting up a link:context-map [context map] and initial interceptors to process that stimuli.
+The service map is processed via the
```
Comment (daemianmack) [inline content/reference/chain-providers.adoc:22]: Is it worth underscoring that _service map_ here is distinct from the _context map_? I could see early readers missing the distinction and getting turned around.

File: content/reference/chain-providers.adoc [hunk 3]
```diff
@@ -8,16 +8,22 @@ Cognitect
...
+link:../api/pedestal.interceptor/io.pedestal.interceptor.chain.html#var-execute[`execute`] function.
+
+Part of a chain provider's responsiblity is to identify when a request has been processed; for
```
Comment (daemianmack) [inline content/reference/chain-providers.adoc:25]: _responsiblity_ => _responsibility_

File: content/reference/chain-providers.adoc [hunk 4]
```diff
@@ -8,16 +8,22 @@ Cognitect
...
+HTTP requests, this is when a handler or some other interceptor attaches a `:response` map to the context map.
+
+Pedestal itself only provide a chain provider implementation for incoming HTTP request processing,
```
Comment (daemianmack) [inline content/reference/chain-providers.adoc:28]: Would suggest...

_Pedestal itself provides a chain provider implementation only for incoming HTTP request processing_

File: content/reference/important-protocols.adoc [hunk 5]
```diff
@@ -40,49 +44,61 @@ reference] for details of their behaviors.
...
+Each of these data structures is a kind of _routing specification_.
+
+In parallel to the data structures, Pedestal has several algorithms for using the routing specification to dispatch
+an incoming request; some algorithms may be most efficient for small services with very few routes, others
+are better for large services with more and more complex routes.
```
Comment (daemianmack) [inline content/reference/important-protocols.adoc:53]: Would suggest _more (and more complex) routes_.

File: content/reference/important-protocols.adoc [hunk 6]
```diff
@@ -40,49 +44,61 @@ reference] for details of their behaviors.
...
+
+The protocols described here are what allows the individual pieces to cooperate; this also means
```
Comment (daemianmack) [inline content/reference/important-protocols.adoc:55]: _are what allows_ => _allow_

File: content/reference/important-protocols.adoc [hunk 7]
```diff
@@ -40,49 +44,61 @@ reference] for details of their behaviors.
...
+that it is possible to define an application-specific routing specification format and use it with a Pedestal-supplied
+algorithm, or using a built-in routing specification and combine it with an application-specific algorithm.
```
Comment (daemianmack) [inline content/reference/important-protocols.adoc:57]: _combine_ => _combining_

**Comment by daemianmack** [comment]
Sniped some micro-suggestions -- hope it's of use. Thanks for improving these docs!

File: content/reference/chain-providers.adoc (hunk 2 above)
Comment (hlship) [reply content/reference/chain-providers.adoc:22]: I think it's ok as it stands, discussion of service map is pretty localized to the parts of the documentation about standing up the service.  I think adding a sidebar or something here will negatively affect readability.

File: content/reference/important-protocols.adoc [hunk 8]
```diff
@@ -24,12 +24,16 @@ Routers use this when building routes: anything that satisfies
...
-`io.pedestal.interceptor/interceptor` function. Queue processing is
-optimized for runtime and does not construct `Interceptor` records.
+[IMPORTANT]
+.*Dynamically added interceptors*
+
+--
+At runtime, new interceptors may be added to the
+link:interceptors#_manipulating_the_interceptor_queue[queue] dynamically;
+(this is what routing itself does); to ensure best runtime performance, `IntoInterceptor` is not
+automatically used; it is assumed any added interceptors are, in fact,
```
Comment (ohpauleez) [inline content/reference/important-protocols.adoc:34]: I don't think I would suggest it's "assumed" here.  In most cases, Pedestal checks chains (and routes) and ensures the IntoIntercator transformation is applied before the service if fully running.

File: content/reference/important-protocols.adoc [hunk 9]
```diff
@@ -40,49 +44,63 @@ reference] for details of their behaviors.
 These protocols break up the lifecycle of routes and allow both
 extension points and alternative strategies in several places.
 
-To give you an idea what that means, there are three built-in route
-definition syntaxes: table, terse, and verbose. There are three built
-in routing algorithms: linear-search, prefix-tree, and map-tree. An
-application can use any syntax with any algorithm. Or, an application
-can create a new syntax while using a built-in algorithm. Or, an
-application could use a built-in syntax with a new algorithm!
+Over time, Pedestal has evolved several data structures for representing routing of HTTP requests;
```
[the remark this answers is not in the document]
Comment (hlship) [reply content/reference/important-protocols.adoc:47]: I'm targeting this documentation at someone who's trying to find their way around Pedestal; like many things in Clojure, it's layered on top of something quite generic, and that's great (we really used this at Walmart for example). I want to ensure that someone coming to these docs to figure out how their HTTP requests get routed aren't overwhelmed by the fact that, technically, the concepts _could_ be adapted to something else; I'm calling that out,, yes, but I want to simplify the documentation towards what 99% of the readers are interested in, 99% of the time.

File: content/reference/important-protocols.adoc (hunk 8 above)
Comment (hlship) [reply content/reference/important-protocols.adoc]: I've converted this to largely just link to the details on queuing interceptors.  Reads better.

File: content/reference/important-protocols.adoc (hunk 9 above)
[the remark this answers is not in the document]
Comment (hlship) [reply content/reference/important-protocols.adoc:39]: Thanks for the feedback!

**Pull request merged**